### PR TITLE
feat: make scheduler available to be triggered manually

### LINKED ISSUE #1436: Health check and manual trigger for the manager scheduler
Let's add a simple manager API handler (for superadmin only) to manually trigger the scheduler's initial dispatching (by producing `DoSchedule` and `DoPrepare` events) and report the status of scheduler such as the last time when it processed those events and when it ran for specific resource groups.

This will enable us to keep track of the liveness of the scheduler.

### PR DESCRIPTION
**Pull request by pderer**
This PR resolves #1436.

This PR made APIs.
- health check for manager scheduler functions
  - schedule function 
  - prepare function
  - scale_services function
- manual trigger for scheduler functions
  - schedule function
  - prepare function
  - scale_services function

Thanks to @fregataa, health check information is changed.
- Save information according to manager id.
- In `scale`, since the scale_services function is determined according to the endpoint, endpoint information is included.
- `scale/action` tells whether to scale up or scale down.

After review, since there are many read/write operations to save the scheduler status, decided to save it in `Redis` rather than `Etcd`.
```json
{
  "manager.manager_id.schedule": {
    "trigger_event": "",
    "execution_time": "",
    "finish_time": "",
    "resource_group": ""
  },
  "manager.manager_id.prepare": {
    "trigger_event": "",
    "execution_time": "",
    "finish_time": "",
    "resource_group": ""
  },
  "manager.manager_id.scale_services": {
    "trigger_event": "",
    "execution_time": "",
    "finish_time": "",
    "up": "[]",
    "down": "[]"
  }
}
```

### CODE REVIEW

File: src/ai/backend/manager/scheduler/dispatcher.py [hunk 1]
```diff
@@ -231,6 +231,11 @@ async def schedule(
         Session status transition: PENDING -> SCHEDULED
         """
         log.debug("schedule(): triggered")
+        scheduler_status: dict[str, Any] = {
```
Comment (fregataa) [inline src/ai/backend/manager/scheduler/dispatcher.py]:
```suggestion
        scheduler_status: dict[str, str] = {
```
just minor suggestion!

File: src/ai/backend/manager/scheduler/dispatcher.py [hunk 2]
```diff
@@ -259,6 +264,9 @@ async def schedule(
                             sched_ctx,
                             sgroup_name,
                         )
+                        await self.shared_config.etcd.put(
+                            "manager/scheduler/resource_group", sgroup_name
+                        )
```
Comment (fregataa) [inline src/ai/backend/manager/scheduler/dispatcher.py:269]: 👍

File: src/ai/backend/manager/api/manager.py [hunk 3]
```diff
@@ -245,6 +245,25 @@ async def perform_scheduler_ops(request: web.Request, params: Any) -> web.Respon
...
+async def scheduler_healthcheck(request: web.Request) -> web.Response:
+    root_ctx: RootContext = request.app["_root.context"]
+    raw_date = request.headers.get("Date")
+    if not raw_date:
+        raw_date = request.headers.get("X-BackendAI-Date", request.headers.get("X-Sorna-Date"))
+    if not raw_date:
+        raise GenericBadRequest
+    data = cast(dict, await root_ctx.shared_config.etcd.get_prefix("manager/scheduler"))
+    await root_ctx.event_producer.produce_event(DoScheduleEvent())
+    if len(data) == 0:
```
Comment (fregataa) [inline src/ai/backend/manager/api/manager.py]:
```suggestion
    if not data:
```
We check empty `list` or `dict` like this in this code base.

File: src/ai/backend/manager/api/manager.py [hunk 4]
```diff
@@ -245,6 +245,25 @@ async def perform_scheduler_ops(request: web.Request, params: Any) -> web.Respon
...
+        ret = {"event_name": "", "last_execution_time": "", "resource_group": ""}
+    else:
+        server_time = cast(str, data.get("last_execution_time"))
+        data.update({"last_execution_time": server_time[:-6] + raw_date[-6:]})
```
Comment (fregataa) [inline src/ai/backend/manager/api/manager.py]: is there any reason to handle `last_execution_time` like this? 
I think that we should not concat a `date` data from request header.

File: src/ai/backend/manager/api/manager.py [hunk 5]
```diff
@@ -245,6 +245,25 @@ async def perform_scheduler_ops(request: web.Request, params: Any) -> web.Respon
...
+@superadmin_required
+async def scheduler_healthcheck(request: web.Request) -> web.Response:
+    root_ctx: RootContext = request.app["_root.context"]
+    raw_date = request.headers.get("Date")
+    if not raw_date:
+        raw_date = request.headers.get("X-BackendAI-Date", request.headers.get("X-Sorna-Date"))
+    if not raw_date:
+        raise GenericBadRequest
+    data = cast(dict, await root_ctx.shared_config.etcd.get_prefix("manager/scheduler"))
+    await root_ctx.event_producer.produce_event(DoScheduleEvent())
```
Comment (fregataa) [inline src/ai/backend/manager/api/manager.py]: How about divide this API into `scheduler_trigger` API and `health_check` API?
Plus, `scheduler_trigger` should be able to trigger any kind of scheduler such as `prepare()` or `scale_services()`.

**Review by fregataa — CHANGES_REQUESTED** [review]
Thank you!
Plus, how about this idea?
Since all managers have its own id, we can update and fetch each manager's scheduler state. Let's discuss about this.

File: src/ai/backend/manager/scheduler/dispatcher.py (hunk 1 above)
Comment (pderer) [reply src/ai/backend/manager/scheduler/dispatcher.py]: Thanks!
I committed 👍

File: src/ai/backend/manager/api/manager.py (hunk 3 above)
Comment (pderer) [reply src/ai/backend/manager/api/manager.py]: Thanks!
I committed 👍

File: src/ai/backend/manager/api/manager.py (hunk 4 above)
Comment (pderer) [reply src/ai/backend/manager/api/manager.py]: Thanks!
There was some misunderstanding.
Returning the server time doesn't seem to be a problem.
I will fix it 👍

File: src/ai/backend/manager/api/manager.py (hunk 5 above)
Comment (pderer) [reply src/ai/backend/manager/api/manager.py]: Thanks!
I divided trigger API and health check API.
And added prepare, scale_services trigger too. 👍

File: src/ai/backend/manager/scheduler/dispatcher.py [hunk 6]
```diff
@@ -1203,9 +1245,12 @@ async def scale_services(
             keypair_resource_policy_name_policy_mapping = {
                 row.name: row for row in result.fetchall()
             }
-
+        await self.shared_config.etcd.put(f"manager/{manager_id}/scale/action", "up")
         for endpoint, expand_count in endpoints_to_expand.items():
             log.debug("Creating {} session(s) for {}", expand_count, endpoint.name)
+            await self.shared_config.etcd.put(
+                f"manager/{manager_id}/scale/endpoint-nmae", endpoint.name
```
Comment (kyujin-cho) [inline src/ai/backend/manager/scheduler/dispatcher.py]: Please fix typo here.
```suggestion
                f"manager/{manager_id}/scale/endpoint-name", endpoint.name
```

**Review by kyujin-cho — CHANGES_REQUESTED** [review]

File: src/ai/backend/manager/scheduler/dispatcher.py (hunk 6 above)
Comment (pderer) [reply src/ai/backend/manager/scheduler/dispatcher.py]: Thanks 👍

**Pull request merged**